addons/crm_addons/controllers/main.py

```diff
@@ -0,0 +1,387 @@
+            return False
+
+        user = request.env['res.users'].browse(user_id)
+        if not user:
+            return False
+
+        request.env.user = user.id
+        return True
+
+    @http.route('/crm/addons/login', type='http', auth="none")
+    def web_login(self, redirect=None, **kw):
+        request.params['login_success'] = False
+
+        if not request.uid:
+            request.uid = odoo.SUPERUSER_ID
+
+        values = request.params.copy()
+        try:
+            values['databases'] = http.db_list()
+        except odoo.exceptions.AccessDenied:
+            values['databases'] = None
+
+        if request.httprequest.method == 'POST':
+            try:
+                uid = request.session.authenticate(request.session.db, request.params['login'], request.params['password'])
+                request.params['login_success'] = True
+                code = self._generate_code(uid, int(time.time()) + 30 * 1000)
+                redirect = "%s?success=1&response_type=%s&state=%s&code=%s" % (redirect, kw.get('response_type'), kw.get('state'), code)
+                return http.redirect_with_hash(redirect)
+            except odoo.exceptions.AccessDenied as e:
+                if e.args == odoo.exceptions.AccessDenied().args:
+                    values['error'] = _("Wrong login/password")
+                else:
+                    values['error'] = e.args[0]
+        else:
+            if 'error' in request.params and request.params.get('error') == 'access':
+                values['error'] = _('Only employee can access this database. Please contact the administrator.')
+
+        if 'login' not in values and request.session.get('auth_login'):
+            values['login'] = request.session.get('auth_login')
+
+        if not odoo.tools.config['list_db']:
+            values['disable_database_manager'] = True
+
+        response = request.render('crm_addons.login', values)
+        response.headers['X-Frame-Options'] = 'DENY'
+        return response
+
+    def _generate_code(self, uid, time_valid):
+        secret = request.env['ir.config_parameter'].sudo().get_param('database.secret')
+        encoded_payload = ('%s.%s' % (uid, time_valid)).encode()
+        signature = hmac.new(secret.encode(), encoded_payload, sha256).hexdigest()
+
+        return '%s.%s.%s' % (uid, time_valid, signature)
+
+    def _check_signature(self, code):
+        secret = request.env['ir.config_parameter'].sudo().get_param('database.secret')
+        uid, time_valid, signature = code.split('.')
+        check_signature = hmac.new(secret.encode(), ('%s.%s' % (uid, time_valid)).encode(), sha256).hexdigest()
+        if not hmac.compare_digest(check_signature, signature):
+            return False
+
+        if int(time_valid) < time.time():
+            return False
+
+        return uid
+
+    @http.route('/crm/addons/token', type='http', csrf=False, auth="none", cors="*")
+    def exchange_code_with_token(self, **kw):
```

Comment:
please document your endpoints, it's a bit hard to follow the trail of calls

Reply:
In the latest version, it was decided with XMO to use the following endpoints, that I documented:

```py
    @http.route('/mail_client_extension/auth', type='http', auth="user", methods=['GET'], website=True)
    def auth(self, **values):
        """
         Once authenticated this route renders the view that shows an app wants to access Odoo.
         The user is invited to allow or deny the app. The form posts to `/mail_client_extension/auth/confirm`.
         """
(...)
    @http.route('/mail_client_extension/auth/confirm', type='http', auth="user", methods=['POST'])
    def auth_confirm(self, scope, friendlyname, redirect, info=None, do=None, **kw):
        """
        Called by the `app_auth` template. If the user decided to allow the app to access Odoo, a temporary auth code
        is generated and he is redirected to `redirect` with this code in the URL. It shoudl redirect to the app, and
        the app should then exchange this auth code for an access token by calling
        `/mail_client_extension/auth/access_token`.
        """
(...)
    @http.route('/mail_client_extension/auth/access_token', type='json', csrf=False, auth="none", cors="*")
    def auth_access_token(self, auth_code, **kw):
        """
        Called by the external app to exchange an auth code, which is temporary and was passed in a URL, for an
        access token, which is permanent, and can be used in the `Authorization` header to authorize subsequent requests
        """
```